sql: support star expansion in sub-expressions

knz opened the pull request:
```sql
SELECT kv.* FROM kv                 -> SELECT k, v FROM kv
SELECT (kv.*) FROM kv               -> SELECT (k, v) FROM kv
SELECT COUNT(DISTINCT kv.*) FROM kv -> SELECT COUNT(DISTINCT (k, v)) FROM kv

-- previously unsupported:
SELECT COUNT(DISTINCT a.*) FROM kv a, kv b
   -> SELECT COUNT(DISTINCT (a.k, a.v)) FROM kv a, kv b
```

For docs: this lifts a previously documented known limitation.

Fixes #15750.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17833)
<!-- Reviewable:end -->

knz committed:
sql: simplify the star detection for CREATE VIEW

Prior to this patch, whether star expansion had occurred or not would
be annotated in every `renderNode`. This patch simplifies the code to
annotate using a single boolean in `planner` instead.

knz committed:
sql: support star expansion in sub-expressions

```sql
SELECT kv.* FROM kv                 -> SELECT k, v FROM kv
SELECT (kv.*) FROM kv               -> SELECT (k, v) FROM kv
SELECT COUNT(DISTINCT kv.*) FROM kv -> SELECT COUNT(DISTINCT (k, v)) FROM kv

-- previously unsupported:
SELECT COUNT(DISTINCT a.*) FROM kv a, kv b
   -> SELECT COUNT(DISTINCT (a.k, a.v)) FROM kv a, kv b
```

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17833#-:-KsEQ71VE2FBc9yF4kIY:bb76hxo)*
<!-- Sent from Reviewable.io -->

knz commented:
TFYR
